File: policies/cyborg/bitworld/among-them/mod-talks/geometry.nim
## Coordinate / camera math.
##
## Pure functions: world↔screen, camera bounds, room and task lookup, and
## the camera↔patch-vote-index conversions used by the localizer's
## patch-hash table.
##
## Phase 1 port from v2:406-504. Notable adaptations:
##
## - `playerWorldX/Y` and `roomName` take `Perception` instead of `Bot` —
##   they read camera state plus a `localized` flag, and that's exactly
##   what `Perception` exposes.
## - The geometry constants (`PlayerScreenX/Y`, `PlayerWorldOffX/Y`)
##   live here because they are derived from screen + sprite geometry and
##   are read by every layer above (frame, localize, actors). Exported.
## - The `<` PatchEntry / `cmpPatchCandidate` comparators stay with the
##   localizer (`localize.nim`) where they are exclusively used.

import protocol
import ../../sim
import types

const
  PlayerScreenX* = ScreenWidth div 2
  PlayerScreenY* = ScreenHeight div 2
  PlayerWorldOffX* = SpriteDrawOffX + PlayerScreenX - SpriteSize div 2
  PlayerWorldOffY* = SpriteDrawOffY + PlayerScreenY - SpriteSize div 2

proc mapIndexSafe*(x, y: int): int =
  ## Returns the map pixel linear index. The "safe" name is preserved
  ## from v2 even though no bounds-check is performed; callers use
  ## `inMap` first when needed.
  y * MapWidth + x

proc minCameraX*(): int =
  ## Smallest possible centred camera X (allows half-screen + sprite
  ## overhang off the left edge).
  -ScreenWidth div 2 - SpriteSize

proc maxCameraX*(): int =
  ## Largest possible centred camera X.
  MapWidth - ScreenWidth div 2 + SpriteSize

proc minCameraY*(): int =
  ## Smallest possible centred camera Y.
  -ScreenHeight div 2 - SpriteSize

proc maxCameraY*(): int =
  ## Largest possible centred camera Y.
  MapHeight - ScreenHeight div 2 + SpriteSize

proc cameraIndex*(x, y: int): int =
  ## Returns the linear vote-table index for one camera offset. Used by
  ## the patch-hash table in `localize`.
  (y - minCameraY()) * (maxCameraX() - minCameraX() + 1) +
    (x - minCameraX())

proc cameraIndexX*(index: int): int =
  ## Returns the camera X coordinate for one vote-table index.
  minCameraX() + index mod (maxCameraX() - minCameraX() + 1)

proc cameraIndexY*(index: int): int =
  ## Returns the camera Y coordinate for one vote-table index.
  minCameraY() + index div (maxCameraX() - minCameraX() + 1)

proc buttonCameraX*(sim: SimServer): int =
  ## Initial camera X guess centred on the emergency button. Used as
  ## the seed for spiral search before the first lock.
  let button = sim.gameMap.button
  clamp(
    button.x + button.w div 2 - PlayerWorldOffX,
    minCameraX(),
    maxCameraX()
  )

proc buttonCameraY*(sim: SimServer): int =
  ## Initial camera Y guess centred on the emergency button.
  let button = sim.gameMap.button
  clamp(
    button.y + button.h div 2 - PlayerWorldOffY,
    minCameraY(),
    maxCameraY()
  )

proc cameraXForWorld*(x: int): int =
  ## Camera X that centres one world X on the player.
  clamp(x - PlayerWorldOffX, minCameraX(), maxCameraX())

proc cameraYForWorld*(y: int): int =
  ## Camera Y that centres one world Y on the player.
  clamp(y - PlayerWorldOffY, minCameraY(), maxCameraY())

proc inMap*(x, y: int): bool =
  ## True when a world pixel is inside the map rectangle.
  x >= 0 and y >= 0 and x < MapWidth and y < MapHeight

proc cameraCanHoldPlayer*(cameraX, cameraY: int): bool =
  ## True when a camera candidate centres a player position that is
  ## inside the map. Localization rejects candidates that don't.
  inMap(cameraX + PlayerWorldOffX, cameraY + PlayerWorldOffY)

proc playerWorldX*(p: Perception): int =
  ## Inferred player collision X coordinate.
  p.cameraX + PlayerWorldOffX

proc playerWorldY*(p: Perception): int =
  ## Inferred player collision Y coordinate.
  p.cameraY + PlayerWorldOffY

proc roomNameAt*(sim: SimServer, x, y: int): string =
  ## Room containing one world point, or "unknown".
  for room in sim.rooms:
    if x >= room.x and x < room.x + room.w and
        y >= room.y and y < room.y + room.h:
      return room.name
  "unknown"

proc roomName*(p: Perception, sim: SimServer): string =
  ## Room containing the inferred player position. Returns "unknown"
  ## when not localized — callers can rely on that to gate room-based
  ## chat / decisions.
  if not p.localized:
    return "unknown"
  let
    px = p.playerWorldX() + CollisionW div 2
    py = p.playerWorldY() + CollisionH div 2
  sim.roomNameAt(px, py)

proc taskCenter*(task: TaskStation): tuple[x: int, y: int] =
  ## Centre pixel for a task station rectangle.
  (task.x + task.w div 2, task.y + task.h div 2)

proc visibleCrewmateWorld*(p: Perception,
                          crewmate: CrewmateMatch): tuple[x: int, y: int] =
  ## Converts one visible crewmate match (screen-coord) to world
  ## coordinates using the current camera lock.
  (p.cameraX + crewmate.x + SpriteDrawOffX,
   p.cameraY + crewmate.y + SpriteDrawOffY)

proc visibleBodyWorld*(p: Perception,
                      body: BodyMatch): tuple[x: int, y: int] =
  ## Converts one visible body match (screen-coord) to world
  ## coordinates.
  (p.cameraX + body.x + SpriteDrawOffX,
   p.cameraY + body.y + SpriteDrawOffY)

# ---------------------------------------------------------------------------
# Central-room helpers (used by the imposter's stuck-detection)
# ---------------------------------------------------------------------------

proc centralRoomCenter*(sim: SimServer): tuple[x: int, y: int] =
  ## Reference point for the central room (the emergency button room).
  let button = sim.gameMap.button
  (button.x + button.w div 2, button.y + button.h div 2)

proc centralRoomName*(sim: SimServer): string =
  ## Room name containing the emergency button.
  let center = sim.centralRoomCenter()
  sim.roomNameAt(center.x, center.y)

proc inCentralRoom*(p: Perception, sim: SimServer): bool =
  ## True when the player is currently inside the central room. Used
  ## by the imposter to detect "stuck in cafeteria" situations.
  if not p.localized:
    return false
  let central = sim.centralRoomName()
  central != "unknown" and p.roomName(sim) == central

File: players/among_them/coborg/perception/geometry.py
"""Camera / world coordinate math. Port of the localize-relevant subset
of ``users/james/personal_cogs/among_them/guided_bot/perception/geometry.nim``.

Two coordinate systems are in play:

- **World** coordinates: pixel positions inside the 952x534 map.
  Tasks, rooms, vents, the emergency button, and the home point all
  live in world coordinates.
- **Screen** coordinates: 128x128 with the player sprite's visual centre
  near (64, 64). The camera ``(cx, cy)`` is the world offset of the
  screen's top-left pixel.

The player's collision-box centre is offset from the camera origin by
``PLAYER_WORLD_OFF_(X|Y)`` because the drawn sprite anchor is shifted
from its hit-box centre.

All functions here are pure: no module state, no allocation. They're
hot in :func:`localize.update_location`'s local-refit path and in
downstream consumers that convert between camera and world coordinates.
"""

from __future__ import annotations

from .data import (
    BUTTON_H,
    BUTTON_W,
    BUTTON_X,
    BUTTON_Y,
    MAP_HEIGHT,
    MAP_WIDTH,
    SPRITE_DRAW_OFF_X,
    SPRITE_DRAW_OFF_Y,
    SPRITE_SIZE,
)
from .frame import SCREEN_HEIGHT, SCREEN_WIDTH

# --- Player sprite anchor (screen) + collision-box offset (world) ---------

# Screen-space anchor of the player sprite (the renderer centres the
# player and scrolls the map under them). Distinct from
# `ignore.PLAYER_SPRITE_ANCHOR_*`, which names the *drawn* sprite anchor
# used to seed the ignore-mask exclusion zone.
PLAYER_SCREEN_X = SCREEN_WIDTH // 2  # 64
PLAYER_SCREEN_Y = SCREEN_HEIGHT // 2  # 64

# World-space offset from the camera origin to the player's inferred
# collision-box centre. `PlayerWorldOffX = SpriteDrawOffX + PlayerScreenX
# - SpriteSize // 2`, verbatim from upstream geometry.nim. With the
# constants we have: 2 + 64 - 6 = 60 (X), 8 + 64 - 6 = 66 (Y).
PLAYER_WORLD_OFF_X = SPRITE_DRAW_OFF_X + PLAYER_SCREEN_X - (SPRITE_SIZE // 2)
PLAYER_WORLD_OFF_Y = SPRITE_DRAW_OFF_Y + PLAYER_SCREEN_Y - (SPRITE_SIZE // 2)


# --- Camera bounds --------------------------------------------------------


def min_camera_x() -> int:
    return -SCREEN_WIDTH // 2 - SPRITE_SIZE


def max_camera_x() -> int:
    return MAP_WIDTH - SCREEN_WIDTH // 2 + SPRITE_SIZE


def min_camera_y() -> int:
    return -SCREEN_HEIGHT // 2 - SPRITE_SIZE


def max_camera_y() -> int:
    return MAP_HEIGHT - SCREEN_HEIGHT // 2 + SPRITE_SIZE


def camera_width() -> int:
    return max_camera_x() - min_camera_x() + 1


def camera_height() -> int:
    return max_camera_y() - min_camera_y() + 1


# --- World ↔ camera helpers ----------------------------------------------


def _clamp(v: int, lo: int, hi: int) -> int:
    return lo if v < lo else hi if v > hi else v


def button_camera_x() -> int:
    """Initial camera X centred on the emergency button. Clamped to the
    camera range so off-edge buttons don't push the seed out of bounds."""
    target = BUTTON_X + BUTTON_W // 2 - PLAYER_WORLD_OFF_X
    return _clamp(target, min_camera_x(), max_camera_x())


def button_camera_y() -> int:
    target = BUTTON_Y + BUTTON_H // 2 - PLAYER_WORLD_OFF_Y
    return _clamp(target, min_camera_y(), max_camera_y())


def camera_x_for_world(world_x: int) -> int:
    """Camera X that puts world ``world_x`` on the player."""
    return _clamp(world_x - PLAYER_WORLD_OFF_X, min_camera_x(), max_camera_x())


def camera_y_for_world(world_y: int) -> int:
    return _clamp(world_y - PLAYER_WORLD_OFF_Y, min_camera_y(), max_camera_y())


def in_map(world_x: int, world_y: int) -> bool:
    return 0 <= world_x < MAP_WIDTH and 0 <= world_y < MAP_HEIGHT


def camera_can_hold_player(cx: int, cy: int) -> bool:
    """True when the camera puts the player's inferred world position
    inside the map rectangle. Used by localize to filter impossible
    candidates the patch-vote step would otherwise score."""
    return in_map(cx + PLAYER_WORLD_OFF_X, cy + PLAYER_WORLD_OFF_Y)


def player_world_x(camera_x: int) -> int:
    return camera_x + PLAYER_WORLD_OFF_X


def player_world_y(camera_y: int) -> int:
    return camera_y + PLAYER_WORLD_OFF_Y
